pkg/tfshim/sdk-v2/provider.go

```diff
@@ -165,7 +164,18 @@ func (p v2Provider) ReadDataDiff(
 	if !ok {
 		return nil, fmt.Errorf("unknown resource %v", t)
 	}
-	diff, err := r.Diff(ctx, nil, configFromShim(c), p.tf.Meta())
+
+	providerOpts, err := getProviderOptions(p.opts)
+	if err != nil {
+		return nil, err
+	}
+	config := configFromShim(c)
+	rawConfig := makeResourceRawConfig(providerOpts.diffStrategy, config, r)
+
+	instanceState := terraform.InstanceState{
```

Comment:
Slightly worrisome that this is not a complete InstanceState, so this is not a complete fix. How does the reference gRPC server in TF SDKv2 handle invokes? 

```
type InstanceState struct {
	// A unique ID for this resource. This is opaque to Terraform
	// and is only meant as a lookup mechanism for the providers.
	ID string `json:"id"`

	// Attributes are basic information about the resource. Any keys here
	// are accessible in variable format within Terraform configurations:
	// ${resourcetype.name.attribute}.
	Attributes map[string]string `json:"attributes"`

	// Ephemeral is used to store any state associated with this instance
	// that is necessary for the Terraform run to complete, but is not
	// persisted to a state file.
	Ephemeral EphemeralState `json:"-"`

	// Meta is a simple K/V map that is persisted to the State but otherwise
	// ignored by Terraform core. It's meant to be used for accounting by
	// external client code. The value here must only contain Go primitives
	// and collections.
	Meta map[string]interface{} `json:"meta"`

	// Tainted is used to mark a resource for recreation.
	Tainted bool `json:"tainted"`

	mu sync.Mutex
}
```

Reply:
It's worth looking at. This seems like it's strictly better, so if emulating upstream perfectly is difficult, then I think we can get the incremental improvement in and then iterate.

Reply:
Yeah I agree!

Reply:
https://github.com/hashicorp/terraform-plugin-sdk/blob/main/helper/schema/grpc_provider.go#L1209 is how TF does it.

Reply:
Notably

```
	diff, err := res.Diff(ctx, nil, config, s.provider.Meta())
	if err != nil {
		resp.Diagnostics = convert.AppendProtoDiag(ctx, resp.Diagnostics, err)
		return resp, nil
	}

	// Not setting RawConfig here is okay, as ResourceData.GetRawConfig()
	// will return a NullVal of the schema if there is no InstanceDiff.
	if diff != nil {
		diff.RawConfig = configVal
	}
```

Reply:
It passes Nil to diff and then it patches RawConfig. I think it might be worth matching this up, it can only take an hour or so.